File: node/fenwicktree/src/types/JSFenwickTree.js
class JSFenwickTree {

    //*****************
    // Class Members **
    //*****************
    #mOriginal;
    #mFenwick;

    //***************
    // Constructor **
    //***************
    constructor(array) { this.#mOriginal = Array.from(array); }

    sum = (index) => {
        let total = 0;
        while(index > 0) {
            total += this.#fenwick()[index - 1];
            index -= this.#lb(index);
        }
        return total;
    }

    range = (from, to) => this.sum(to) - this.sum(from - 1);

    update = (index, value) => {
        const diff = value - this.#fenwick()[index - 1];
        while(index < this.#fenwick().length) {
            this.#fenwick()[index - 1] += diff;
            index += this.#lb(index);
        }
    }

    //*******************
    // Private Members **
    //*******************

    #createFenwick = () => {
        const array = Array.from(this.#original());
        array.forEach((next, index) => {
            const i = ++index + this.#lb(index);
            if(i < array.length + 1) {
                array[i - 1] += next;
            }
        });
        return array;
    }

    #lb = (value) => value & (-value);

    #fenwick = () => this.#mFenwick === undefined ? this.#mFenwick = this.#createFenwick() : this.#mFenwick;

    #original = () => this.#mOriginal;
}

module.exports = JSFenwickTree;

File: java/fenwicktree/src/main/java/com/js/datastructuresalgorithms/fenwicktree/JSFenwickTree.java
package com.js.datastructuresalgorithms.fenwicktree;

import java.util.Arrays;

public class JSFenwickTree {

    //*****************
    // Class Members **
    //*****************
    private final int[] mOrginal;
    private int[] mFenwick;

    //***************
    // Constructor **
    //***************
    public JSFenwickTree(int[] orginal) { mOrginal = orginal; }

    public int sum(int to) { return sumArray(to); }

    public int range(int begin, int end) { return sumArray(end) - sumArray(begin - 1); }

    public void update(int to, int value) {
        int diff = value - fenwick()[to];
        while(to < fenwick().length) {
            fenwick()[to - 1] += diff;
            to += +lsb(to);
        }
    }

    //*******************
    // Private Members **
    //*******************

    private int sumArray(int index)
    {
        int total = 0;
        while (index > 0) {
            total += fenwick()[index - 1];
            index -= lsb(index);
        }
        return total;
    }

    private int[] original()
    {
        return mOrginal;
    }

    private int[] fenwick()
    {
        return mFenwick == null ? mFenwick = createFenwick() : mFenwick;
    }

    private int[] createFenwick()
    {
        mFenwick = Arrays.copyOf(original(), original().length);
        for (int i = 0; i < mFenwick.length; i++) {
            int index = i + lsb(i + 1);
            if(index < mFenwick.length) {
                mFenwick[index] += mFenwick[i];
            }
        }

        return mFenwick;
    }

    private int lsb(int num) { return num & (-num); }
}
